app/server-utils/url-to-iframe.spec.ts

```diff
@@ -0,0 +1,25 @@
+import {urlToIframe} from './url-to-iframe'
+
+describe('urlToIframe', () => {
+  describe('should not convert a markdown link', () => {
+    it('whose text does not match url', () => {
+      const input = `[this map](https://aisafety.world/)`
+      expect(urlToIframe(input)).toBe(input)
+    })
+
+    it('whose text matches url but is not whitelisted', () => {
+      const url = 'https://example.com/'
+      const input = `[${url}](${url})</a>`
+      expect(urlToIframe(input)).toBe(input)
+    })
+  })
+
+  describe('should convert a markdown link', () => {
+    it('whose text matches url and is whitelisted', () => {
```

Comment:
could you also test these?:
* `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)`
* `[https://aisafety.world/](https://aisafety.world/?bla=bla)`
* `[https://aisafety.world/?bla=bla](https://aisafety.world/)`
* `[http://aisafety.world/](http://aisafety.world/)`
* `[aisafety.world](http://aisafety.world/)`
* `[aisafety.world/](aisafety.world/)`

Reply:
Yes, certainly. Thanks for thinking of all these cases.

What should the expected results be for each of them in your view? Should all of them convert into the iframe or just some of them?

I went back and looked at the GH issue for this iframe (https://github.com/StampyAI/stampy-ui/issues/140) and @plexish was saying that "If a full plain URL from the whitelisted domains is in the markdown, it should be transformed into a large (~screen height, full answer width) iframe of that URL."
Re-reading that, I wonder if my initial understanding and implementation isn't quite correct. The current implementation is considering the `href`/`()` portion of the link but maybe this is not relevant. Maybe it just matters that the text/`[]` portion is a whitelisted URL.

With this new understanding, I think the expected behaviour should be:
1. `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/?bla=bla`
1. `[https://aisafety.world/](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/`
1. `[https://aisafety.world/?bla=bla](https://aisafety.world/)`  render iframe with href of `https://aisafety.world/?bla=bla`
1. `[http://aisafety.world/](http://aisafety.world/)` -> should render iframe with href of `https://aisafety.world/`
1. `[aisafety.world](http://aisafety.world/)` -> should **not** render iframe because it's missing URI scheme
1. `[aisafety.world/](aisafety.world/)`  -> should **not** render iframe because it's missing URI scheme

Does the above seem right?

Reply:
I'd personally go with both having to be identical. But what you propose is also totally reasonable. The main thing is to have a consistent view

Reply:
I'm okay with either of these schemes a long as it's documented somewhere the editors can read like the formatting question.

Reply:
Ok, I think "both having to be indentical" is the easier route for now and I ideally don't want to change the implementation in an "adding tests" PR.

That said, the first test case you suggested doesn't create an iframe with the current implementation. I think the query string breaks it.
> `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/?bla=bla`

I made it a failing test case: https://github.com/StampyAI/stampy-ui/pull/193/commits/97390534ac31e9f464066b639997ae36285b96e0
We can come back and add support for query strings in a separate PR, IMO

Reply:
nice! @Aprillion would you prefer this to be fixed now, or do you think that urls with query strings should work as they do now? 
@jrhender if you want to fix it, the regex should be `const matches = text.matchAll(/\[(https?:\/\/[\.\w]+?\/?[^\]]*?)\]\(\1\)/gi)`, as it is in the comment

Reply:
I would prefer to fix it, that's the purpose of tests, no strong opinion between following options:
* a) this PR with [it.failing(...)](https://jestjs.io/docs/api#testfailingname-fn-timeout), and the PR author can file a GH issue with the bug and then someone can fix code in next PR, or
* b) 2 separate commits in this PR so that tests are passing before we merge, or
* c) 1 commit with a passing test and fixed code at the same time

just no merging of red tests, please

up to @jrhender what option you feel most comfortable with, though TBH, option b) sounds most straight forward (if the code from above comment fixes the problem without making the other tests unhappy)

Reply:
Ah I didn't now about `it.failing`, that's nifty.

Ok, I moved querystring case to be expected to pass and adjusted regex as per mruwnik's tips https://github.com/StampyAI/stampy-ui/pull/193/commits/566a78fc4f18e2ffde2246e583222f8022f7298c

Reply:
> as it's documented somewhere the editors can read like the formatting question.

@plexish I'm glad to document this. Do you know where a good place to do this might be? Is there a Q&A writing guide perhaps?

Reply:
yep write it in a comment there and I'll include it @jrhender